### PR TITLE
[train/tune] Load the latest unflattened metrics with `Result.from_path`

### LINKED ISSUE #40535: [Air] Ray Result object not properly deserialized
### What happened + What you expected to happen

Ray automatically saves a metrics to a `result.json` file, and users can reinstantiate results with `Result.from_path(<PATH TO result.json)`. However, during deserialization in `from_path`, the metrics are read with [pd.json_normalize(json_list, sep="/")](https://github.com/ray-project/ray/blob/e19b0eb1cb898f4f38349783fd8f86cbc8c60b83/python/ray/air/result.py#L156), then the `latest_metrics` are built by taking the [last row of that dataframe and using the `to_dict` method](https://github.com/ray-project/ray/blob/e19b0eb1cb898f4f38349783fd8f86cbc8c60b83/python/ray/air/result.py#L166). It's that `latest_metrics` object which is passed to the `metrics` kwarg to create the `Result` object that is returned.

The result of the combined `sep="/"` and `to_dict()` call is that the metrics end up collapsing nested dicts into `/`-separated strings. In particular, this means that the config (which is considered part of the metrics) is collapsed into `{"config/key_1": "val_1", "config/key_2": "val_2"}`.

Then when we use the `Result.config` property to fetch the config from the result, it [tries to do](https://github.com/ray-project/ray/blob/e19b0eb1cb898f4f38349783fd8f86cbc8c60b83/python/ray/air/result.py#L67) `self.metrics.get("config")`, except there is no more `config` key, only the `config/foo` nested keys.

I believe the fix for this is relatively simple and if someone gives the go-ahead, I'm happy to create a PR:

[Current implementation](https://github.com/ray-project/ray/blob/e19b0eb1cb898f4f38349783fd8f86cbc8c60b83/python/ray/air/result.py#L153-L156)
```python
if result_json_file.exists():
    with open(result_json_file, "r") as f:
        json_list = [json.loads(line) for line in f if line]
        metrics_df = pd.json_normalize(json_list, sep="/")

# (a few lines later)
latest_metrics = metrics_df.iloc[-1].to_dict() if not metrics_df.empty else {}
```

Proposed implementation
```python
if result_json_file.exists():
    with open(result_json_file, "r") as f:
        json_list = [json.loads(line) for line in f if line]
        metrics_df = pd.json_normalize(json_list, sep="/")
        # use the parsed json dict directly
        latest_metrics = json_list[-1]
```

### Versions / Dependencies

Python 3.10.x
Ray 2.7.1

### Reproduction script

This can be reproduced with any ray `Result` object:

```python
result_1 = <some ray result>
result_2 = Result.from_path(result_1.path)
assert result_1.config is not None
assert result_2.config is None
```

(I used [this example](https://docs.ray.io/en/latest/train/examples/pytorch/torch_fashion_mnist_example.html#torch-fashion-mnist-ex) to verify.)

Screen Shot 2023-10-20 at 3 36 39 PM

### Issue Severity

Low: It annoys or frustrates me.

### PR DESCRIPTION
**Pull request by justinvyu**
## Why are these changes needed?

A trial's `config` gets saved as part of an iteration's reported metric. It gets saved by the default `JsonLoggerCallback` in each row of `result.json` as an unflattened dict.

The deserialization logic in `Result.from_path` reads this `result.json` file to reconstruct the config, but it flattens the dict along the way, result in keys like `"config/a"`, leading to `"config"` not being found.

This PR fixes this by loading the metrics from json without doing extra operations on it.

cc: @daturkel

### Limitations

`progress.csv` is another auto-generated file created by the default `CSVLoggerCallback`. There is no need for both of these files -- we may want to just only save the `result.json` in the future. One difference is that `progress.csv` does not contain the config as part of each row. In the case that a user disables the JSON logger callback, or deletes the `result.json` file, `Result.from_path` falls back to load info from this csv file, which doesn't include config information. `result.config` would return `None` in this case.

## Related issue number

Closes https://github.com/ray-project/ray/issues/40535

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by woshiyyya — APPROVED** [review]
Thank you @justinvyu! LGTM

One small question is, the keys in `latest_metrics` and `metrics_df` will be mismatched. Can we update the docstring of class `Result` and provide an simple example?
